### ISSUE TITLE
Literature pointers for better understanding the `Cnn14_DecisionLevelAtt` model

### ISSUE THREAD

**Issue by sirgarfieldc** [qiuqiangkong/audioset_tagging_cnn#32]
Hello, Thanks for the awesome repo.

I am new to Audio & SED domain. I have been using your arch for one of the recent Kaggle competition and getting decent result. Therefore, I would like to better understand details of `Cnn14_DecisionLevelAtt` 

I have read the [PANNs paper](https://arxiv.org/abs/1912.10211), but it mostly focuses on the CNN feature extractor part. I am interested in understanding why things are done in the way they are for the `Cnn14_DecisionLevelAtt` model ( basically everything beside the CNN feature extractor ).  Can you point me to some write-ups that explains this ?

Thanks

**Comment by qiuqiangkong** [comment]
Hi, see this paper: https://arxiv.org/pdf/1710.00343 for the DecisionLevelAtt model. For audio tagging, it is suggested to use Cnn14. For SED, in our later experiments we find DecisionLevelMax also works well, sometimes even better than DecisionLevelAtt.

**Comment by sirgarfieldc** [comment]
Thanks for pointers !

**Comment by sirgarfieldc** [comment]
Hello, I just read the referenced paper. A couple of follow-ups.

Q1:

For audio tagging with weak label task,  which backbone architecture would you recommend ?  PANNs or the CRNNs + GLU model proposed in [reference paper](https://arxiv.org/pdf/1710.00343 ).  

Q2.
 
For `DecisionLevelMax`, I have being using max aggregation to convert framewise_output to clipwise_output. I notice that the [reference paper](https://arxiv.org/pdf/1710.00343 ) says it should be averaging. For audio tagging with weak label set-up, do you think that I am making a mistake in using the max ? Or is the max / average choice task dependent and should be tuned as hyper-parameter instead ?

**Comment by qiuqiangkong** [comment]
Hi,

    For Q1, for audio tagging, PANNs are later work, and achieves better
performance than the CRNNs + GLU system. So it is suggested to use PANNs,
such as the Cnn14 is a good choice.

    For Q2, DecisionLevelMax is suggested for sound event detection (SED),
but not audio tagging. Cnn14 system can achieve better performance than the
DecisionLevel models. For SED, the max operation in DecisionLevelMax can
reduce the number of false alarms (meanwhile missing some
events). DecisionLevelAvg and DecisionLevelAtt also works, and may achieve
better F1 score for sound event detection (not evaluated on AudioSet).
However, for real applications we usually want to avoid false alarm,
so DecisionLevelMax works better in this case.

**Comment by sirgarfieldc** [comment]
Hello, I have been using your arch for this [Kaggle Competition](https://www.kaggle.com/c/rfcx-species-audio-detection/leaderboard). Now, I am at top3% and I am trying to make a modification to your architecture ( unsuccessfully) to get to top 1%.  😄 .  I am wondering if you are willing to provide some insights and suggestion to my failed modification.

## Background info:

I am using your `PANNs DecisionLevelAttn` to perform audio tagging on 60 secs long audio clips. With the interested events being extremely short ( average 1-2 secs). The objective is to do multi-class classification on 24 labels.

I notice that frequency is a good discriminator for the labels because events from each label is clustered around certain unique frequency band.

## Intended modification idea:

```
 x = torch.mean(x, dim=3)

```
In my hypothesis, this step is throwing away a lot of useful info, as we just blindly mean aggregate away all the frequency info.
My idea is to apply some sort of simple linear attention and use a different aggregation scheme to obtain a unique feature representation to predict each label ( so that not useful frequency could be ignored ).  

## Assistance requested:

Unfortunately, my modification did not work out well. I am wondering if you have any tips on how you would modify the model to take advantage of the task specific insight I just described ?   I have summarized my modification below, and if you have any insight on what mistake I am making. It would also be very helpful. 


## My modification attempt:



#### Change 1

In __init__, I construct a unique linear attention layer (for frequency aggregation) and attention block to predict each label.
here attention size `args.attn_size`  is 1280 // num_lables , I am trying to prevent over parameterization. 

In code, it is

```
self.att_block = [ AttBlock(args.attn_size, 1, activation='sigmoid').cuda() for i in range(24) ]         
self.attention_heads = [ nn.Linear(11, 1, bias=True).cuda() for i in range(24)]

```

### Change 2

After the cnn feature extraction step, I send the representation to `label_specific_forward` and then concat each label prediction in the end.

```
 x = self.backbone.extract_features(x)

 o = [ self.label_specific_forward(x, idx, frames_num) for idx in range(24) ]

  clipwise_output = [item[0] for item in o]
  logit = [item[1] for item in o]
  framewise_output = [item[2] for item in o]

  clipwise_output = torch.cat(clipwise_output, dim=-1)
  logit = torch.cat(logit, dim=-1)
  framewise_output = torch.cat(framewise_output, dim=-1)

```

### Change 3:

Inside the `label_specific_forward`, I use label specific frequency aggregation and label specific attention block 
```
x_freq_agg = self.attention_heads[idx](cnn_feature).squeeze(dim=-1)
 
# Copy past original implementation ...

(clipwise_output, norm_att, segmentwise_output) = self.att_block[idx](x)
```

**Comment by qiuqiangkong** [comment]
Hi TacoKoala,

    This looks an interesting challenge! I looked into this task, and found
the task is *audio tagging*, but not sound event detection. The Cnn14
architecture can perform better for audio tagging (mAP=0.43 on AudioSet)
compared to the DecisionLevelAtt system (mAP=0.39 on AudioSet). I am sure
the Cnn14 architecture will give you some surprise!

For change 1, one option is to use flatten operation to replace the mean
operation.

For change 2, for this kaggle task, framewise_output is not necessary. So
architectures such as Cnn14 can directly predict the clip-level probability.

For change 3, I was not successful using frequency domain attention in
audio tagging tasks. But I thought flatten frequency + fully connected
layer should work well.

P.S. Did you load the pretrained weights from PANNs? Load all weights and
finetune all weights will give the best performance.

Best wishes,

Qiuqiang

**Comment by sirgarfieldc** [comment]
1) Will try the CNN14 arch. Thanks for the suggestion

2) Will try out the flatten + fully connected suggestion 

3) PANNs's backbone is not effective for this particular task ( general consensus from multiple participants of the competition). I am using efficientnet-b4 as backbone

4) Empirically, max of framewise_output is a lot better than clipwise_output ( with the same model weightl) for this task. The difference could be as large as 6% on leader board. ( The competition is ranked based precision, where probability does not matter only the relative ranking ) My intuition is that maybe framewise_output is better at detecting short event ? ( Since most event of interest is 1-2 secs )

**Comment by qiuqiangkong** [comment]
Hi TacoKoala,

    Thanks for the observations! This may indicate that the data is very
weakly labelled, that is the sound event only appear a very short time in
the 60-second audio clip. In this case, removing the .mean() from .mean() +
.max() in the global pooling layer might help?

Best wishes,

Qiuqiang

**Comment by sirgarfieldc** [comment]
I have a additional question regarding this detail 

```
x = x.transpose(1, 3)
x = self.bn0(x)
x = x.transpose(1, 3)

```

We are using bn layer to learn the best frequency axis normalization transform instead of just using the computed stats from the training data.  Does this trick have a name in the literature ?  Does it typically work better than just using the fixed stats ? 

Also why do we just stop at freq axis only, why don't we do it over time axis as well ? 
EX:

```
x = x.transpose(1, 3)
x = self.freq_bn0(x)
x = x.transpose(1, 3)

x = x.transpose(1, 2)
x = self.time_bn0(x)
x = x.transpose(1, 2)

```

**Comment by qiuqiangkong** [comment]
Hi TacoKoala,

    In the past, people usually normalize each channel (frequency bins),
including substract mean value and divide the standard value. Notice the
mean and standard value is calculated over the whole dataset. We use bn to
replace operation, and there is no name for this yet. In the past, the
normalization is only done in the frequency domain, but not the time
domain. Hope it helps!

Best wishes,

Qiuqiang

**Comment by ruclion** [comment]
help a lot！ Try to follow~
